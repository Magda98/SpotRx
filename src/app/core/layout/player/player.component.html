@if(player) {
<div class="player">
  <ng-container>
    @if(playerService.playbackState | async; as position){
    <mat-slider class="song-slider" min="1" [max]="player.duration" step="10">
      <input (dragEnd)="setPosition($event)" [value]="position" matSliderThumb>
    </mat-slider>
    }
    <div class="song-info">
      <img [alt]="player.track_window.current_track.album.name" class="album-image"
        [src]="player.track_window.current_track.album.images[0].url">
      <div class="album-artist">
        <h2 class="name">{{player.track_window.current_track.name}}</h2>
        <span class="album-name">
          {{player.track_window.current_track.album.name}}
        </span>
        <div class="artist-container">
          @for(item of player.track_window.current_track.artists; track item.name){
          <span class="artist">{{item.name}}</span>
          }
        </div>
      </div>
    </div>
    <div class="action-buttons-container">
      <div class="action-buttons">
        <button appButton type="button" aria-label="previous" [mode]="'text'" (click)="playerService.skipPrev()">
          <mat-icon>skip_previous</mat-icon>
        </button>
        <button appButton type="button" aria-label="play" [mode]="'text'" (click)="playerService.tooglePlay()">
          @if(player.paused){
          <mat-icon>play_circle_outline</mat-icon>
          }
          @else{
          <mat-icon>pause_circle_outline</mat-icon>
          }
        </button>
        <button appButton type="button" aria-label="next" [mode]="'text'" (click)="playerService.skipNext();">
          <mat-icon>skip_next</mat-icon>
        </button>
      </div>
    </div>
    <div class="slider-container">
      <button appButton type="button" [mode]="'text'">
        <app-icon [ngClass]="{'primary': player.shuffle }" (click)="toggleShuffle()" class="icon"
          type="shuffle"></app-icon>
      </button>
      <button appButton type="button" [mode]="'text'" (click)="toggleSavedTrack()">
        <app-icon [ngClass]="{'primary': isSavedTrack() === true }" class="icon" type="heart"></app-icon>
      </button>
      <mat-slider class="volume-slider" min="0" max="1" step="0.1">
        <input (dragEnd)="setVolume($event)" value="0.5" matSliderThumb>
      </mat-slider>
    </div>
  </ng-container>
</div>
}