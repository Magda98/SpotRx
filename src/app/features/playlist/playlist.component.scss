.playlist {
	display: flex;
	flex-direction: column;
	justify-content: center;
	max-height: 80vh;

	.title {
		margin-bottom: 32px;
		text-align: center;
	}
}
