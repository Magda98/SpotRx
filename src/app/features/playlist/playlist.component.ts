import { ActivatedRoute } from '@angular/router';
import { Component, inject, signal } from '@angular/core';
import { distinctUntilChanged, map, tap } from 'rxjs';
import { PageEvent } from '@angular/material/paginator';
import { toSignal } from '@angular/core/rxjs-interop';
import { injectQuery } from '@tanstack/angular-query-experimental';
import { NgxSkeletonLoaderModule } from 'ngx-skeleton-loader';
import { TrackListComponent } from '@app/shared/components/track-list/track-list.component';
import { TrackService } from '@app/shared/services/track.service';

@Component({
	selector: 'app-playlist',
	templateUrl: './playlist.component.html',
	styleUrls: ['./playlist.component.scss'],
	imports: [TrackListComponent, NgxSkeletonLoaderModule],
})
export class PlaylistComponent {
	private route = inject(ActivatedRoute);
	private trackService = inject(TrackService);
	offset = signal(0);
	playlistId = toSignal(
		this.route.paramMap.pipe(
			map((paramMap) => paramMap.get('id')),
			distinctUntilChanged(),
			tap(() => {
				this.offset.set(0);
			}),
		),
	);
	tracksQuery = injectQuery(() =>
		this.trackService.getPlaylistTracks(this.playlistId() ?? '', this.offset()),
	);
	playlistInfo = injectQuery(() => this.trackService.getPlaylistInfo(this.playlistId() ?? ''));

	getNextPage(page: PageEvent) {
		this.offset.set(page.pageSize * page.pageIndex);
	}
}
