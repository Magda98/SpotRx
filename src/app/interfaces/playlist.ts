import { Image } from "./image"

export interface Playlist {
    href: string,
    id: string,
    images: Image[],
    name: string,
    type: string,
    uri: string,
}


export interface PlaylistResponse {
    items: Playlist[],
    total: number,
    
}