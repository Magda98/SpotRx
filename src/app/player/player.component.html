<div class="player">
    <!-- <mat-slider min="1" max="5" step="0.5" value="1.5"></mat-slider> -->
    <button *ngIf="playerState | async as player" class="play" (click)="togglePlay()">
        <mat-icon *ngIf="player.paused">play_circle_outline</mat-icon>
        <mat-icon *ngIf="!player.paused">pause_circle_outline</mat-icon>
        {{player.position}}{{player.paused}}
    </button>

</div>