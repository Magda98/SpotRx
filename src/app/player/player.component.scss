.player {
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 140px;
  background-color: #2c2f31;
  padding-left: 0;
  box-sizing: border-box;
  display: grid;
  grid-template-columns: 1fr;
  padding-top: 16px;

  @media screen and (min-width: 980px) {
    grid-template-columns: 1fr 1fr 1fr;
    padding-left: 220px;
  }

  .song-slider {
    position: absolute;
    top: -23px;
    left: 0;
    right: 0;

    @media screen and (min-width: 980px) {
      left: 212px;
    }
  }

  .action-buttons {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-left: 0;
    padding: 0 8px;

    @media screen and (min-width: 980px) {
      margin-left: 25%;
      width: 100%;
      max-width: 250px;
    }
  }

  .song-info {
    display: flex;
    align-items: center;
    padding: 0 8px;

    .album-image {
      width: 60px;
      height: 60px;
      object-fit: cover;
      border-radius: 5px;
      display: block;
      margin-right: 10px;

      @media screen and (min-width: 980px) {
        width: 80px;
        height: 80px;
      }
    }

    .name {
      margin-bottom: 0;
    }

    .artist-container {
      display: flex;
      opacity: 0.6;
      margin-top: 5px;
    }

    .artist {
      margin-right: 5px;
    }

    .album-artist {
      display: flex;
      flex-direction: column;
    }

    .album-name {
      opacity: 0.6;
      font-size: 12px;

      @media screen and (min-width: 980px) {
        font-size: 16px;
      }
    }
  }

  .slider-container {
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .volume-slider {
    display: none;

    @media screen and (min-width: 980px) {
      display: block;
      max-width: 150px;
      width: 100%;
    }
  }
}
