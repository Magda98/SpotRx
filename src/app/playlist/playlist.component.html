<div class="playlist">
  @if(playlistInfo.isSuccess()){
  <h1 class="title">{{playlistInfo.data().name}}</h1>
  }
  @if(playlistInfo.isLoading()){
  <div class="title">
    <ngx-skeleton-loader class="artist-skeleton" animation="progress-dark" count="1"
      [theme]="{ height: '32px', width: '200px', 'background-color': '#2c2f31' }"></ngx-skeleton-loader>
  </div>
  }
  <app-track-list class="track-list" (getNextPage)="getNextPage($event)" [total]="playlistInfo.data()?.tracks?.total"
    [tracksListQuery]="tracksQuery">
  </app-track-list>
</div>