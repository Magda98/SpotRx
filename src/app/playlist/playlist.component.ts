import { TrackService } from '../services/track.service';
import { ActivatedRoute } from '@angular/router';
import { Component, inject, signal } from '@angular/core';
import { distinctUntilChanged, map, tap } from 'rxjs';
import { PageEvent } from '@angular/material/paginator';
import { toSignal } from '@angular/core/rxjs-interop';
import { injectQuery } from '@tanstack/angular-query-experimental';
import { TrackListComponent } from '../shared/track-list/track-list.component';
import { NgxSkeletonLoaderModule } from 'ngx-skeleton-loader';

@Component({
  selector: 'app-playlist',
  templateUrl: './playlist.component.html',
  styleUrls: ['./playlist.component.scss'],
  standalone: true,
  imports: [TrackListComponent, NgxSkeletonLoaderModule],
})
export class PlaylistComponent {
  private route = inject(ActivatedRoute);
  private trackService = inject(TrackService);
  offset = signal(0);
  playlistId = toSignal(
    this.route.paramMap.pipe(
      map((paramMap) => paramMap.get('id')),
      distinctUntilChanged(),
      tap(() => {
        this.offset.set(0);
      })
    )
  );
  tracksQuery = injectQuery(() =>
    this.trackService.getPlaylistTracks(this.playlistId() ?? '', this.offset())
  );
  playlistInfo = injectQuery(() =>
    this.trackService.getPlaylistInfo(this.playlistId() ?? '')
  );

  getNextPage(page: PageEvent) {
    this.offset.set(page.pageSize * page.pageIndex);
  }
}
