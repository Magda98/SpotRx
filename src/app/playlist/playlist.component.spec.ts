import {
  render,
  screen,
  waitForElementToBeRemoved,
} from '@testing-library/angular';
import { PlaylistComponent } from './playlist.component';
import {
  QueryClient,
  provideAngularQuery,
} from '@tanstack/angular-query-experimental';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { MockBackendInterceptor } from '../../tests/mock-backend.interceptor';
import { ActivatedRoute, convertToParamMap } from '@angular/router';
import { of } from 'rxjs';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { PlayerService } from '../services/player.service';

describe('PlaylistComponent', () => {
  test('should render', async () => {
    const component = await render(PlaylistComponent, {
      imports: [HttpClientTestingModule],
      componentProviders: [
        {
          provide: ActivatedRoute,
          useValue: {
            paramMap: of(convertToParamMap({ id: '50Gsv3p7qLLPVzfPBu8UcO' })),
          },
        },
      ],
      providers: [
        provideAngularQuery(new QueryClient()),
        {
          provide: HTTP_INTERCEPTORS,
          useClass: MockBackendInterceptor,
          multi: true,
        },
        PlayerService,
      ],
    });

    await waitForElementToBeRemoved(await screen.findAllByRole('progressbar'));
    expect(component.container).toMatchSnapshot();
  });
});
