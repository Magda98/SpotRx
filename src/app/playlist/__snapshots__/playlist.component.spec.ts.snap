// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaylistComponent should render 1`] = `
<div
  id="root0"
>
  <div
    class="playlist"
  >
    <h1
      class="title"
    >
      Dua Lipa üçëüí´üî•
    </h1>
    
    
    <app-track-list
      class="track-list"
    >
      <div
        class="track-list-container"
      >
        <mat-paginator
          aria-label="Select page"
          class="mat-mdc-paginator paginator"
          role="group"
        >
          <div
            class="mat-mdc-paginator-outer-container"
          >
            <div
              class="mat-mdc-paginator-container"
            >
              <div
                class="mat-mdc-paginator-page-size"
              >
                <div
                  class="mat-mdc-paginator-page-size-label"
                >
                   Items per page: 
                </div>
                
                <div
                  class="mat-mdc-paginator-page-size-value"
                >
                  6
                </div>
                
              </div>
              
              <div
                class="mat-mdc-paginator-range-actions"
              >
                <div
                  aria-live="polite"
                  class="mat-mdc-paginator-range-label"
                >
                   1 ‚Äì 1 of 1 
                </div>
                
                <button
                  aria-label="Previous page"
                  class="mat-mdc-tooltip-trigger mat-mdc-paginator-navigation-previous mdc-icon-button mat-mdc-icon-button mat-mdc-button-disabled _mat-animation-noopable mat-unthemed mat-mdc-button-base mat-mdc-tooltip-disabled"
                  disabled="true"
                  mat-icon-button=""
                  mat-ripple-loader-centered=""
                  mat-ripple-loader-class-name="mat-mdc-button-ripple"
                  mat-ripple-loader-disabled=""
                  mat-ripple-loader-uninitialized=""
                  type="button"
                >
                  <span
                    class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"
                  />
                  <svg
                    aria-hidden="true"
                    class="mat-mdc-paginator-icon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                    />
                  </svg>
                  <span
                    class="mat-mdc-focus-indicator"
                  />
                  <span
                    class="mat-mdc-button-touch-target"
                  />
                </button>
                
                <button
                  aria-label="Next page"
                  class="mat-mdc-tooltip-trigger mat-mdc-paginator-navigation-next mdc-icon-button mat-mdc-icon-button mat-mdc-button-disabled _mat-animation-noopable mat-unthemed mat-mdc-button-base mat-mdc-tooltip-disabled"
                  disabled="true"
                  mat-icon-button=""
                  mat-ripple-loader-centered=""
                  mat-ripple-loader-class-name="mat-mdc-button-ripple"
                  mat-ripple-loader-disabled=""
                  mat-ripple-loader-uninitialized=""
                  type="button"
                >
                  <span
                    class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"
                  />
                  <svg
                    aria-hidden="true"
                    class="mat-mdc-paginator-icon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                    />
                  </svg>
                  <span
                    class="mat-mdc-focus-indicator"
                  />
                  <span
                    class="mat-mdc-button-touch-target"
                  />
                </button>
                
                
              </div>
            </div>
          </div>
        </mat-paginator>
        <ul
          class="track-list"
        >
          <li
            class="track"
          >
            <div
              class="song-info"
            >
              <img
                class="track-img"
                fetchpriority="auto"
                height="90"
                loading="lazy"
                ng-img="true"
                src="https://i.scdn.co/image/ab67616d00001e02d1410c1372fab1e516328fa8"
                width="90"
              />
              <div
                class="album-artist"
              >
                <h2
                  class="name"
                >
                  Santa's Coming for Us
                </h2>
                <span
                  class="album-name"
                >
                   Everyday Is Christmas (Deluxe Edition) 
                </span>
                <div
                  class="track-artist"
                >
                  <span
                    class="artist"
                  >
                    Sia
                  </span>
                  
                </div>
              </div>
            </div>
            <span
              class="time"
            >
              3:26 
            </span>
          </li>
          
        </ul>
        
        
      </div>
    </app-track-list>
  </div>
</div>
`;
